### PR TITLE
Refactoring dependency projectId to be stored as groupId and artifactId 

### PR DESCRIPTION
**Pull request by gs-gunjan**

### CODE REVIEW

File: legend-sdlc-model/src/main/java/org/finos/legend/sdlc/domain/model/project/configuration/ProjectDependency.java [hunk 1]
```diff
@@ -98,7 +98,7 @@ public static ProjectDependency parseProjectDependency(String string, int start,
             throw new IllegalArgumentException("Invalid project dependency string: null");
         }
 
-        int delimiterIndex = string.indexOf(delimiter, start);
+        int delimiterIndex = string.lastIndexOf(delimiter);
         if ((delimiterIndex == -1) || (delimiterIndex >= end))
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-model/src/main/java/org/finos/legend/sdlc/domain/model/project/configuration/ProjectDependency.java:102]: Make sure that you're keeping within the start/end bounds.

```suggestion
        int delimiterIndex = string.lastIndexOf(delimiter, end - 1);
        if ((delimiterIndex == -1) || (delimiterIndex < start))
```

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectStructure.java [hunk 2]
```diff
@@ -446,6 +446,19 @@ else if (requireRevisionId)
...
+        Set<ProjectDependency> toUpdateProjectDependencies = projectDependencies.stream().filter(dep -> !dep.getProjectId().contains(":")).collect(Collectors.toSet());
+
+        if (toUpdateProjectDependencies.size() > 0)
+        {
+            for (ProjectDependency projectDependency : toUpdateProjectDependencies)
+            {
+                updateProjectDependencies = true;
+                ProjectConfiguration dependencyConfig = getProjectConfiguration(projectFileAccessProvider.getFileAccessContext(projectDependency.getProjectId(),projectDependency.getVersionId()));
+                projectDependencies.remove(projectDependency);
+                projectDependencies.add(ProjectDependency.newProjectDependency(dependencyConfig.getGroupId() + ":" + dependencyConfig.getArtifactId(),projectDependency.getVersionId()));
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectStructure.java:458]: You need some null checking here. First, dependencyConfig could be null. Even if it's not, the group or artifact ids (or both) could be null.

**Review by kevin-m-knight-gs — CHANGES_REQUESTED** [review]

File: legend-sdlc-server/src/test/java/org/finos/legend/sdlc/server/project/TestProjectStructure.java [hunk 3]
```diff
@@ -281,6 +281,21 @@ protected void testUpdateGroupAndArtifactIds(ProjectType projectType)
...
+        Assert.assertEquals(GROUP_ID, mavenCoordinate[0]);
+        Assert.assertEquals("testproject0", mavenCoordinate[1]);
+        Assert.assertEquals("0.0.1", projectDependency.getVersionId().toVersionIdString());
+
+        //backward compatibility
+        ProjectDependency oldProjectDependency = ProjectDependency.parseProjectDependency("testproject:0.0.1");
+        Assert.assertEquals("testproject", oldProjectDependency.getProjectId());
+        Assert.assertEquals("0.0.1", oldProjectDependency.getVersionId().toVersionIdString());
+    }
+
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/test/java/org/finos/legend/sdlc/server/project/TestProjectStructure.java:298]: This belongs in TestProjectDependency, not in TestProjectStructure.

File: legend-sdlc-server/src/test/java/org/finos/legend/sdlc/server/project/TestProjectStructure.java [hunk 4]
```diff
@@ -296,13 +311,13 @@ public void testUpdateProjectDependencies_Prototype()
     protected void testUpdateProjectDependencies(ProjectType projectType)
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/test/java/org/finos/legend/sdlc/server/project/TestProjectStructure.java:311]: Could you make sure that some test (either this one or a new one) demonstrates upgrading old-style dependencies to new-style?

**Review by kevin-m-knight-gs — CHANGES_REQUESTED** [review]

File: legend-sdlc-server/src/test/java/org/finos/legend/sdlc/server/project/TestProjectStructure.java [hunk 5]
```diff
@@ -44,6 +46,7 @@
 import org.junit.Test;
 
 import java.time.Instant;
+import java.util.ArrayList;
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/test/java/org/finos/legend/sdlc/server/project/TestProjectStructure.java:49]: Is this import still used?

File: legend-sdlc-server/src/test/java/org/finos/legend/sdlc/server/project/TestProjectStructure.java [hunk 6]
```diff
@@ -58,6 +61,8 @@
 import java.util.function.Function;
 import java.util.stream.Collectors;
 
+import static org.finos.legend.sdlc.server.project.ProjectStructure.PROJECT_CONFIG_PATH;
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/test/java/org/finos/legend/sdlc/server/project/TestProjectStructure.java:64]: Can we avoid using a static import for this?

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectStructure.java [hunk 7]
```diff
@@ -725,6 +686,63 @@ else if (!Arrays.equals(currentBytes, newBytes))
...
+    private static void updateOrAddDependencies(Set<ProjectDependency> toUpdateProjectDependencies, ProjectFileAccessProvider projectFileAccessProvider, Set<ProjectDependency> projectDependencies, boolean purgeOldDependency)
+    {
+        List<ProjectDependency> unknownDependencies = Lists.mutable.empty();
+        List<ProjectDependency> nonProdDependencies = Lists.mutable.empty();
+        SortedMap<ProjectDependency, Exception> accessExceptions = new TreeMap<>();
+        for (ProjectDependency projectDependency : toUpdateProjectDependencies)
+        {
+            try
+            {
+                ProjectConfiguration dependencyConfig = getProjectConfiguration(projectFileAccessProvider.getFileAccessContext(projectDependency.getProjectId(), projectDependency.getVersionId()));
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectStructure.java:698]: What if projectDependency has a new-style project id (i.e., groupId:artifactId)? Won't this fail?

**Review by kevin-m-knight-gs — CHANGES_REQUESTED** [review]

File: legend-sdlc-model/src/main/java/org/finos/legend/sdlc/domain/model/project/configuration/ProjectDependency.java [hunk 8]
```diff
@@ -72,6 +72,11 @@ public StringBuilder appendVersionIdString(StringBuilder builder)
         return (versionId == null) ? builder.append("null") : versionId.appendVersionIdString(builder);
     }
 
+    public static boolean isLegacyProjectDependency(ProjectDependency projectDependency)
+    {
+        return !projectDependency.getProjectId().contains(":");
```
[the remark this answers is not in the document]
Comment (gs-gunjan) [reply legend-sdlc-model/src/main/java/org/finos/legend/sdlc/domain/model/project/configuration/ProjectDependency.java:77]: Will throw an exception now, if the project id is null

File: legend-sdlc-server/src/test/java/org/finos/legend/sdlc/server/project/TestProjectStructure.java (hunk 6 above)
Comment (gs-gunjan) [reply legend-sdlc-server/src/test/java/org/finos/legend/sdlc/server/project/TestProjectStructure.java:64]: yes will do

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectStructureV11Factory.java [hunk 9]
```diff
@@ -345,33 +346,37 @@ public void addVersionPackageModulePlugins(String name, BiFunction<String, Versi
...
+                                builder.append(" (").append(dependencyConfig.getGroupId()).append(':').append(dependencyConfig.getArtifactId()).append(':');
+                                projectDependency.getVersionId().appendVersionIdString(builder);
+                                builder.append(')');
+                            }
                         }
                     }
+                    throw new LegendSDLCServerException(builder.toString());
                 }
-                throw new LegendSDLCServerException(builder.toString());
             }
```
[the remark this answers is not in the document]
Comment (gs-gunjan) [reply legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectStructureV11Factory.java]: deleting this extra check

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/maven/MavenProjectStructure.java [hunk 10]
```diff
@@ -275,10 +275,15 @@ public void addJacksonDependency(Consumer<Dependency> dependencyConsumer, boolea
 
     public static Stream<Dependency> projectDependencyToMavenDependencies(ProjectDependency projectDependency, BiFunction<String, VersionId, ProjectFileAccessProvider.FileAccessContext> versionFileAccessContextProvider, Collection<? extends ArtifactType> artifactTypes, boolean setVersion)
     {
+        String versionString = setVersion ? projectDependency.getVersionId().toVersionIdString() : null;
+        if (!ProjectDependency.isLegacyProjectDependency(projectDependency))
+        {
+            String[] mavenCoordinates = projectDependency.getProjectId().split(":");
+            return artifactTypes.stream().map(artifactType -> newMavenDependency(mavenCoordinates[0], mavenCoordinates[1] + "-" + artifactType, versionString));
```
[the remark this answers is not in the document]
Comment (gs-gunjan) [reply legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/maven/MavenProjectStructure.java]: Doing a double check for artifactTypes and also created a new function for which parses the new projectIds

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/maven/MavenProjectStructure.java [hunk 11]
```diff
@@ -275,10 +276,16 @@ public void addJacksonDependency(Consumer<Dependency> dependencyConsumer, boolea
 
     public static Stream<Dependency> projectDependencyToMavenDependencies(ProjectDependency projectDependency, BiFunction<String, VersionId, ProjectFileAccessProvider.FileAccessContext> versionFileAccessContextProvider, Collection<? extends ArtifactType> artifactTypes, boolean setVersion)
     {
+        String versionString = setVersion ? projectDependency.getVersionId().toVersionIdString() : null;
+        if (!ProjectDependency.isLegacyProjectDependency(projectDependency))
+        {
+            List<String> mavenCoordinates = ProjectDependency.getMavenCoordinatesFromProjectDependency(projectDependency.getProjectId());
+            List<ArtifactType> artifactTypeList = ((artifactTypes == null) || artifactTypes.isEmpty()) ? Arrays.asList(ArtifactType.entities,ArtifactType.versioned_entities) : artifactTypes.stream().collect(Collectors.toList());
+            return artifactTypeList.stream().map(artifactType -> newMavenDependency(mavenCoordinates.get(0), mavenCoordinates.get(1) + "-" + artifactType, versionString));
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/maven/MavenProjectStructure.java:284]: The default set of artifact types should be a static field. E.g.,

`private static final ImmutableList<ArtifactType> DEFAULT_ARTIFACT_TYPES = Lists.immutable.with(ArtifactType.entities, ArtifactType.versioned_entities);`

Then this bit of code should be something like this:
```suggestion
            Collection<? extends ArtifactType> resolvedArtifactTypes = ((artifactTypes == null) || artifactTypes.isEmpty()) ? DEFAULT_ARTIFACT_TYPES : artifactTypes;
            return resolvedArtifactTypes.stream().map(artifactType -> newMavenDependency(mavenCoordinates.get(0), mavenCoordinates.get(1) + "-" + artifactType, versionString));
```

**Review by kevin-m-knight-gs — CHANGES_REQUESTED** [review]

**Review by kevin-m-knight-gs — APPROVED** [review]

**Pull request merged**